### PR TITLE
[SEMI-MODULAR] PIXEL-SHIFTING module

### PR DESCRIPTION
**Pull request by Azarak**
## About The Pull Request
Adds pixel shifting into the game, enabling you to smoothly shift around on the tile you're in. With a few changes.
Movement is now smooth, you do not need to spam the key you wish to move in, you may hold it down and watch yourself smoothly move.
Keybind is now ALT.
Rebinds face direction to CTRL.

## Why It's Good For The Game

No, don't hurt me! 

## Changelog
:cl: Azarak, Gandalf2k15
add: Added pixelshifting. (Default keybinding is ALT, check your keybinds!)
/:cl:

### CODE REVIEW

**Review by Gandalf2k15 — APPROVED** [review]

**Comment by Azarak** [comment]
I cant figure out the issue that makes the keybinds do other variants if one isn't executed successfully

**Comment by Gandalf2k15** [comment]
Fixed this.

File: code/modules/mob/mob_movement.dm [hunk 1]
```diff
@@ -89,6 +89,17 @@
...
+	//SKYRAT EDIT ADDITION BEGIN - PIXEL_SHIFT
+	if(mob.shifting)
+		mob.pixel_shift(direct)
+		return FALSE
+	else
+		if(mob.is_shifted)
+			if(isliving(mob))
+				mob.unpixel_shift(TRUE)
+			else
+				mob.unpixel_shift(FALSE)
```
Comment (Azarak) [inline code/modules/mob/mob_movement.dm]:
```suggestion
	if(isliving(mob))
		var/mob/living/M = mob
		if(M.shifting)
			M.pixel_shift(direct)
			return FALSE
		else if(M.is_shifted)
			pixel_x = M.get_standard_pixel_x_offset(!M.mobility_flags & MOBILITY_STAND)
			pixel_y = M.get_standard_pixel_y_offset(!M.mobility_flags & MOBILITY_STAND)
```

File: modular_skyrat/modules/pixel_shift/code/pixel_shift.dm [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+/mob/proc/unpixel_shift(living)
+	if(is_shifted)
+		is_shifted = FALSE
+		if(living)
+			var/mob/living/M = src
+			pixel_x = M.get_standard_pixel_x_offset(!M.mobility_flags & MOBILITY_STAND)
+			pixel_y = M.get_standard_pixel_y_offset(!M.mobility_flags & MOBILITY_STAND)
+		else
+			pixel_x = 0
+			pixel_y = 0
```
Comment (Azarak) [inline modular_skyrat/modules/pixel_shift/code/pixel_shift.dm]:
```suggestion
```

File: modular_skyrat/modules/pixel_shift/code/pixel_shift.dm [hunk 3]
```diff
@@ -0,0 +1,66 @@
+/mob
```
Comment (Azarak) [inline modular_skyrat/modules/pixel_shift/code/pixel_shift.dm:1]:
```suggestion
/mob/living
```

File: modular_skyrat/modules/pixel_shift/code/pixel_shift.dm [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+/datum/keybinding/mob/pixel_shift/up(client/user)
+	. = ..()
+	if(.)
+		return
+	var/mob/M = user.mob
+	M.shifting = FALSE
+	return TRUE
+
+/mob/proc/pixel_shift(direction)
```
Comment (Azarak) [inline modular_skyrat/modules/pixel_shift/code/pixel_shift.dm:39]:
```suggestion
/mob/living/proc/pixel_shift(direction)
```

File: modular_skyrat/modules/pixel_shift/code/pixel_shift.dm [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+/datum/keybinding/mob/pixel_shift/down(client/user)
+	. = ..()
+	if(.)
+		return
+	var/mob/M = user.mob
+	M.shifting = TRUE
+	return TRUE
+
+/datum/keybinding/mob/pixel_shift/up(client/user)
```
Comment (Azarak) [inline modular_skyrat/modules/pixel_shift/code/pixel_shift.dm:22]:
```suggestion
/datum/keybinding/living/pixel_shift/up(client/user)
```

File: modular_skyrat/modules/pixel_shift/code/pixel_shift.dm [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+/datum/keybinding/mob/pixel_shift
+	hotkey_keys = list("Ctrl")
+	name = "pixel_shift"
+	full_name = "Pixel Shift"
+	description = "Shift your characters offset."
+	category = CATEGORY_MOVEMENT
+	keybind_signal = COMSIG_KB_MOB_PIXELSHIFT
+
+/datum/keybinding/mob/pixel_shift/down(client/user)
```
Comment (Azarak) [inline modular_skyrat/modules/pixel_shift/code/pixel_shift.dm:14]:
```suggestion
/datum/keybinding/living/pixel_shift/down(client/user)
```

File: modular_skyrat/modules/pixel_shift/code/pixel_shift.dm [hunk 7]
```diff
@@ -0,0 +1,55 @@
+/mob/living
+	///Whether the mob is pixel shifted or not
+	var/is_shifted
+	var/shifting //If we are in the shifting setting.
+
+/datum/keybinding/mob/pixel_shift
```
Comment (Azarak) [inline modular_skyrat/modules/pixel_shift/code/pixel_shift.dm:6]:
```suggestion
/datum/keybinding/living/pixel_shift
```

File: modular_skyrat/modules/pixel_shift/code/pixel_shift.dm [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+	full_name = "Pixel Shift"
+	description = "Shift your characters offset."
+	category = CATEGORY_MOVEMENT
+	keybind_signal = COMSIG_KB_MOB_PIXELSHIFT
+
+/datum/keybinding/mob/pixel_shift/down(client/user)
+	. = ..()
+	if(.)
+		return
+	var/mob/M = user.mob
```
Comment (Azarak) [inline modular_skyrat/modules/pixel_shift/code/pixel_shift.dm:18]:
```suggestion
	var/mob/living/M = user.mob
```

File: modular_skyrat/modules/pixel_shift/code/pixel_shift.dm [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+	M.shifting = TRUE
+	return TRUE
+
+/datum/keybinding/mob/pixel_shift/up(client/user)
+	. = ..()
+	if(.)
+		return
+	var/mob/M = user.mob
```
Comment (Azarak) [inline modular_skyrat/modules/pixel_shift/code/pixel_shift.dm:26]:
```suggestion
	var/mob/living/M = user.mob
```

File: modular_skyrat/modules/pixel_shift/code/pixel_shift.dm [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+	M.shifting = FALSE
+	return TRUE
+
+/mob/proc/unpixel_shift()
```
Comment (Azarak) [inline modular_skyrat/modules/pixel_shift/code/pixel_shift.dm:30]:
```suggestion
/mob/proc/unpixel_shift()
	return

/mob/living/unpixel_shift()
```

File: modular_skyrat/modules/pixel_shift/code/pixel_shift.dm [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+		is_shifted = FALSE
+		if(isliving(src))
+			var/mob/living/M = src
+			pixel_x = M.get_standard_pixel_x_offset(!M.mobility_flags & MOBILITY_STAND)
+			pixel_y = M.get_standard_pixel_y_offset(!M.mobility_flags & MOBILITY_STAND)
+		else
+			pixel_x = 0
+			pixel_y = 0
+
+/mob/proc/pixel_shift(direction)
```
Comment (Azarak) [inline modular_skyrat/modules/pixel_shift/code/pixel_shift.dm:39]:
```suggestion
/mob/proc/pixel_shift(direction)
	return

/mob/living/pixel_shift(direction)
```

File: modular_skyrat/modules/pixel_shift/code/pixel_shift.dm [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+/mob/proc/unpixel_shift()
+	if(is_shifted)
+		is_shifted = FALSE
+		if(isliving(src))
+			var/mob/living/M = src
+			pixel_x = M.get_standard_pixel_x_offset(!M.mobility_flags & MOBILITY_STAND)
+			pixel_y = M.get_standard_pixel_y_offset(!M.mobility_flags & MOBILITY_STAND)
+		else
+			pixel_x = 0
+			pixel_y = 0
```
Comment (Azarak) [inline modular_skyrat/modules/pixel_shift/code/pixel_shift.dm]:
```suggestion
		pixel_x = M.get_standard_pixel_x_offset(!M.mobility_flags & MOBILITY_STAND)
		pixel_y = M.get_standard_pixel_y_offset(!M.mobility_flags & MOBILITY_STAND)
```

**Review by Azarak — COMMENTED** [review]
Thoughts on this to remove a type check?

File: modular_skyrat/modules/pixel_shift/code/pixel_shift.dm [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+	return TRUE
+
+/mob/proc/unpixel_shift()
+	return
+
+/mob/living/unpixel_shift()
+	if(is_shifted)
+		is_shifted = FALSE
+		pixel_x = M.get_standard_pixel_x_offset(!M.mobility_flags & MOBILITY_STAND)
+		pixel_y = M.get_standard_pixel_y_offset(!M.mobility_flags & MOBILITY_STAND)
```
Comment (Azarak) [inline modular_skyrat/modules/pixel_shift/code/pixel_shift.dm]:
```suggestion
		pixel_x = get_standard_pixel_x_offset(!M.mobility_flags & MOBILITY_STAND)
		pixel_y = get_standard_pixel_y_offset(!M.mobility_flags & MOBILITY_STAND)
```

**Comment by Azarak** [comment]
Was tested, should be fine, dont wanna wait for travis

**Pull request merged**